kad-dht/README.md

```diff
@@ -0,0 +1,386 @@
+	FindPeer(context.Context, peer.ID) (pstore.PeerInfo, error)
+}
+
+// ValueStore is a basic Put/Get interface.
+type ValueStore interface {
+	// PutValue adds value corresponding to given Key.
+	PutValue(context.Context, string, []byte, ...ropts.Option) error
+
+	// GetValue searches for the value corresponding to given Key.
+	GetValue(context.Context, string, ...ropts.Option) ([]byte, error)
+}
+```
+
+## Value lookups
+
+When looking up an entry in the DHT, the implementor should collect at least `Q`
+(quorum) responses from distinct nodes to check for consistency before returning
+an answer.
+
+Should the responses be different, the `Validator.Select()` function is used to
+resolve the conflict and select the _best_ result.
+
+**Entry correction.** Nodes that returned _worse_ records are updated via a
+direct `PUT_VALUE` RPC call when the lookup completes. Thus the DHT network
+eventually converges to the best value for each record, as a result of nodes
+collaborating with one another.
+
+### Algorithm
+
+Let's assume we’re looking for key `K`. We first try to fetch the value from the local store. If found, and `Q == { 0, 1 }`, the search is complete.
+
+Otherwise, the local result counts for one towards the search of `Q` values. We then enter an iterative network search.
+
+We keep track of:
+
+* the number of values we've fetched (`cnt`).
+* the best value we've found (`best`), and which peers returned it (`Pb`)
+* the set of peers we've already queried (`Pq`) and the set of next query candidates sorted by distance from `K` in ascending order (`Pn`).
+* the set of peers with outdated values (`Po`). 
+
+**Initialization**: seed `Pn` with the `α` peers from our routing table we know are closest to `K`, based on the XOR distance function.
+
+**Then we loop:**
+
+*WIP (raulk): lookup timeout.*
+
+1. If we have collected `Q` or more answers, we cancel outstanding requests, return `best`, and we notify the peers holding an outdated value (`Po`) of the best value we discovered, by sending `PUT_VALUE(K, best)` messages. _Return._
+2. Pick as many peers from the candidate peers (`Pn`) as the `α` concurrency factor allows. Send each a `GET_VALUE(K)` request, and mark it as _queried_ in `Pq`.
+3. Upon a response:
+	1. If successful, and we receive a value:
+		1. If this is the first value we've seen, we store it in `best`, along
+           with the peer who sent it in `Pb`.
+		2. Otherwise, we resolve the conflict by calling `Validator.Select(best,
+           new)`:
+			1. If the new value wins, store it in `best`, and mark all formerly
+               “best" peers (`Pb`) as _outdated peers_ (`Po`). The current peer
+               becomes the new best peer (`Pb`).
+			2. If the new value loses, we add the current peer to `Po`.
+	2. If successful without a value, the response will contain the closest
+       nodes the peer knows to the key `K`. Add them to the candidate list `Pn`,
+       except for those that have already been queried.
```

Comment:
```suggestion
	2. If successful with or without a value, the response will contain the closest
       nodes the peer knows to the key `K`. Add them to the candidate list `Pn`,
       except for those that have already been queried.
```

As the specification is currently worded, if (for example) `Q > α` and all of the first `α` queries return a value, then `Pn` will be empty, but `Q` answers won't yet have been collected.

Reply:
Thanks @tmakarios for pointing this out.

https://github.com/libp2p/specs/pull/108/commits/0126af14220d2935b6d0a173e65e3d73273f122a adjusts the specification accordingly.

For the record, both go-libp2p and rust-libp2p sent `closerPeers` detached from whether they sent `record`.

https://github.com/libp2p/go-libp2p-kad-dht/blob/eac1b5e1ad7a90ffee7a7d4b0650e04a621fbb84/handlers.go#L55-L91

https://github.com/libp2p/rust-libp2p/blob/c5bcada2c2a1e34b8e39b885f10ed7c2d00e8373/protocols/kad/src/behaviour.rs#L1779-L1804